iroh/src/magicsock.rs

```diff
@@ -563,16 +566,16 @@ impl MagicSock {
                         // at any time so these errors should be treated as transient and
                         // are just timeouts.  Hence we opt for returning Ok.  See
                         // test_try_send_no_udp_addr_or_relay_url to explore this further.
-                        error!(
+                        debug!(
```

Comment:
This one I find tricky to judge.  I'm not entirely sure if this is a benign condition.

Reply:
It happens every time I run a test.

I think the condition this happens in is for example: I create a new `Endpoint` with relays disabled.
Then I send data with the udp socket send eventually returning `pending`.

Reply:
If you look at the code above:
- `udp_pending` ends up being `true`
- `relay_pending` is false, because `relay_url` is `None`
- `relay_sent` and `udp_sent` are both `false`
- `pings_sent` ends up being `false` too, maybe because there's just no pings scheduled to send.

Reply:
Huh, in that case is it easy to create an isolated reproducer for this?

Reply:
> If you look at the code above:
> 
>     * `udp_pending` ends up being `true`
> 
>     * `relay_pending` is false, because `relay_url` is `None`
> 
>     * `relay_sent` and `udp_sent` are both `false`
> 
>     * `pings_sent` ends up being `false` too, maybe because there's just no pings scheduled to send.

Nice!

So this means this **is** a bug: the socket should return `WouldBlock` in this case.  Right?

Reply:
> So this means this **is** a bug: the socket should return `WouldBlock` in this case. Right?

Not sure. There's another comment close by that says:

> Returning WouldBlock triggers a hot loop

Reply:
To summarise an out-of-band conversation: this can not return WouldBlock because the next datagram being sent may have a destination with a RelayUrl so if you would go into blocking here you'd be wrongly blocking.

I've been sufficiently convinced now that this really is a normal situation and that we can turn down the log level here.

Could you add a comment explaining when this happens though?  Basically some version of https://github.com/n0-computer/iroh/pull/3016#discussion_r1873229444 maybe with the bit that you can't block because you can't predict the dest of the next datagram.

Reply:
I added more comments. Lmk what you think :)